### PR TITLE
Add documentation command

### PR DESCRIPTION
**Pull request by adi88d**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/20939

## Description
Add `generate-docs` command to demisto-sdk for generate documentation automatically to integrations, playbooks and scripts.

### CODE REVIEW

File: demisto_sdk/main.py [hunk 1]
```diff
@@ -468,6 +471,35 @@ def init(**kwargs):
...
+    "-id", "--id_set", help="Path of updated id_set.json file.", required=False)
+@click.option(
+    "-v", "--verbose", is_flag=True, help="Verbose output - mainly for debugging purposes")
+def generate_doc(file_type, **kwargs):
+    if file_type == 'integration':
+        return generate_integration_doc(**kwargs)
+    elif file_type == 'script':
+        return generate_script_doc(**kwargs)
+    elif file_type == 'playbook':
+        return generate_playbook_doc(**kwargs)
```
Comment (yaakovi) [inline demisto_sdk/main.py]:
```
else:
    print_error(...)
    return 1
```

File: demisto_sdk/generate_docs/generate_script_doc.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    id_set_sections.remove('TestPlaybooks')
+
+    for key in id_set_sections:
+        items = id_set[key]
+        for item in items:
+            key = list(item.keys())[0]
+            scripts = item[key].get('implementing_scripts', [])
+            if scripts and script_id in scripts:
+                used_in_list.append(item[key].get('name', []))
+    return list(filter(None, dict.fromkeys(used_in_list)))
```
Comment (IkaDemisto) [inline demisto_sdk/generate_docs/generate_script_doc.py]: Why do you need it?

File: demisto_sdk/runners/runner.py [hunk 3]
```diff
@@ -122,3 +122,29 @@ def _export_debug_log(self, log_ids: list):
...
+        update_entry = {
+            'investigationId': playground_id,
+            'data': command
+        }
+        res = self.client.investigation_add_entries_sync(update_entry=update_entry)
+
+        body = {'query': '${.}'}
+        context = self.client.generic_request(f'investigation/{playground_id}/context', 'POST', body)[0]
+
+        import ast
```
Comment (IkaDemisto) [inline demisto_sdk/runners/runner.py]: Just move the import to top of the module.

**Comment by yaakov** [comment]
Wrong Ya’akov, please remove me.

Thanks.

File: demisto_sdk/generate_docs/common.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+       :return: A modified string
+       :rtype: ``str``
+    """
+    if escape_multiline:
+        st = st.replace('\r\n', '<br>')  # Windows
+        st = st.replace('\r', '<br>')  # old Mac
+        st = st.replace('\n', '<br>')  # Unix
+
+    if minimal_escaping:
+        for c in '|':
```
[the remark this answers is not in the document]
Comment (adi88d) [reply demisto_sdk/generate_docs/common.py]: to replace any chars that might break a markdown string

File: demisto_sdk/generate_docs/generate_integration_doc.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            if cmd not in examples:
+                examples[cmd] = (example, md_example, context_example)
+    return examples, errors
+
+
+def run_command(command_example):
+    errors = []
+    context_example = ''
+    md_example = ''
+    cmd = command_example
```
[the remark this answers is not in the document]
Comment (adi88d) [reply demisto_sdk/generate_docs/generate_integration_doc.py]: command_example is a string

File: demisto_sdk/generate_docs/generate_integration_doc.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    Note: if a command appears more then once, run all occurrences but stores only the first.
+    """
+    examples = {}  # type: dict
+    errors = []  # type: list
+    for example in command_examples:
+        if example.startswith('!'):
+            cmd, md_example, context_example, cmd_errors = run_command(example)
+            errors.extend(cmd_errors)
+
+            if cmd not in examples:
```
[the remark this answers is not in the document]
Comment (adi88d) [reply demisto_sdk/generate_docs/generate_integration_doc.py]: if a command appears more then once in command_examples

File: demisto_sdk/generate_docs/generate_integration_doc.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+        docs.extend(generate_section('Known Limitations', ''))
+        # Troubleshooting
+        docs.extend(generate_section('Troubleshooting', ''))
+
+        doc_text = '\n'.join(docs)
+
+        save_output(output, 'README.md', doc_text)
+
+        if errors:
+            print_error('Possible Errors:')
```
[the remark this answers is not in the document]
Comment (adi88d) [reply demisto_sdk/generate_docs/generate_integration_doc.py]: print_error method print in new line

File: demisto_sdk/generate_docs/generate_playbook_doc.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    commands = []
+    playbooks = []
+
+    playbook_tasks = playbook.get('tasks')
+    for task in playbook_tasks:
+        task = playbook_tasks[task]['task']
+        if task['iscommand']:
+            integration = task['script']
+            integration = integration.split('|||')
+            integrations.append(integration[0])
```
[the remark this answers is not in the document]
Comment (adi88d) [reply demisto_sdk/generate_docs/generate_playbook_doc.py]: string with the name of the integration, this command for example return 'gmail':
`script: Gmail|||gmail-search`

File: demisto_sdk/generate_docs/generate_playbook_doc.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    :param playbook: the playbook object.
+    :return: list of inputs and list of errors.
+    """
+    errors = []
+    inputs = []
+
+    if not playbook.get('inputs'):
+        return {}, []
+
+    for _input in playbook.get('inputs'):
```
[the remark this answers is not in the document]
Comment (adi88d) [reply demisto_sdk/generate_docs/generate_playbook_doc.py]: For not shadowing built in names

File: demisto_sdk/generate_docs/generate_script_doc.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+            script = get_yaml(input)
+            script_py = script.get('script', '-')
+            if script_py == '-':
+                print_error(f'Error: script code not found in {input}')
+                return
+            script_yml = input
+
+        doc = []
+        errors = []
+        script = get_yaml(script_yml)
```
[the remark this answers is not in the document]
Comment (adi88d) [reply demisto_sdk/generate_docs/generate_script_doc.py]: It contains the errors from the documentation process, used in lines 67,68 and 71

File: demisto_sdk/generate_docs/generate_script_doc.py (hunk 2 above)
Comment (adi88d) [reply demisto_sdk/generate_docs/generate_script_doc.py]: To delete duplicates from the list, i replaced it to set.

File: demisto_sdk/main.py (hunk 1 above)
Comment (adi88d) [reply demisto_sdk/main.py]: BTW if the type is not the the list of the options, you gets this error:
`Error: Invalid value for "-t" / "--file_type": invalid choice: widget. (choose from integration, script, playbook)
`

File: demisto_sdk/generate_docs/common.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (IkaDemisto) [reply demisto_sdk/generate_docs/common.py]: ? How many times are you entering the loop here?

**Comment by glicht** [comment]
@adi88d please fix the lgtm comments about import *.

**Comment by glicht** [comment]
Does this run on a package dir? Or do I need to give it a path to a unified yml?

**Comment by adi88d** [comment]
@glicht the input is yml file, it doesnt checks the code so the file can be unified or not.

**Comment by glicht** [comment]
Tried running this against ConvertFile

Couldn't get it to work for executing an example. Got a very bad error message which doesn't help out. Ran the following:
```bash
demisto-sdk generate-docs -i Scripts/ConvertFile/ConvertFile.yml -o Scripts/ConvertFile -e '!ConvertFile entry_id=${File.EntryID} format=pdf'
```
Result:
```
Start generate script documentation...
2020-01-30 13:51:13,451 WARNING Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1056)'))': /investigations/search
WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1056)'))': /investigations/search
2020-01-30 13:51:13,468 WARNING Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1056)'))': /investigations/search
WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1056)'))': /investigations/search
2020-01-30 13:51:13,483 WARNING Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1056)'))': /investigations/search
WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1056)'))': /investigations/search
Error: 'ConvertFile'
```
Screenshot:
image

When running without an example I got. Following output:
```
Start generate script documentation...
Output file was saved to :
Scripts/ConvertFile/README.md
Possible Errors:
Script example is missing.
id_set.json file is missing
```
Screen shot:
image

Not clear why we are printing the message about id_set.json. What is a user supposed to do with this?

Change the message about the example missing to be more clear. Use the following:

Note: Script example was not provided. For a more complete documentation run with the -e options and specify how to run the script. For example: `-e '!ConvertFile entry_id=<entry_id>'

Also for the generated README.md I have the following comments:
* List of sections at start should be removed (we will have this anyway in the table of contents in the site)
* `This is the metadata of the script.` text doesn't provide any value. Talk with Alex about this or remove it.
* Docker Image: remove this. i don't want to maintain this field as it will become outdated.
 
image

**Comment by adi88d** [comment]
@glicht The problem with the script example fixed,
i'll talk with @kirbles19 tomorrow about the README format

File: demisto_sdk/commands/generate_docs/common.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    Generate simple section in markdown format.
+    :param title: The section title.
+    :param data: The section text.
+    :return: array of strings contains the section lines in markdown format.
+    """
+    section = [
+        '## {}'.format(title),
+        ''
+    ]
+    if data:
```
Comment (IkaDemisto) [inline demisto_sdk/commands/generate_docs/common.py]: If data is not mandatory, put it as optional parameter

File: demisto_sdk/commands/generate_docs/common.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+     :return: array of strings contains the list section in markdown format.
+     """
+    section = []
+    if title:
+        section.append(f'{header_type} {title}')
+
+    if horizontal_rule:
+        section.append('---')
+
+    if not data:
```
Comment (IkaDemisto) [inline demisto_sdk/commands/generate_docs/common.py]: Same here

File: demisto_sdk/commands/generate_docs/common.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+    :return: array of strings contains the table in markdown format.
+    """
+    section = []
+    if title:
+        section.append(f'## {title}')
+
+    if horizontal_rule:
+        section.append('---')
+
+    if not data:
```
Comment (IkaDemisto) [inline demisto_sdk/commands/generate_docs/common.py]: And here

**Pull request merged**